eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/ServiceClientInstantiationCheck.java

```diff
@@ -0,0 +1,167 @@
+     *  Log error if the service client has public or protected constructor.
+     *
+     * @param token the CTOR_DEF AST node
+     */
+    private void checkNonPublicOrProtectedConstructor(DetailAST token) {
+        DetailAST modifiersToken = token.findFirstToken(TokenTypes.MODIFIERS);
+        AccessModifier accessModifier = CheckUtil.getAccessModifierFromModifiersToken(modifiersToken);
+        if (accessModifier.equals(AccessModifier.PUBLIC) || accessModifier.equals(AccessModifier.PROTECTED)) {
+            log(modifiersToken, "Subclass of ServiceClient should not have any public or protected constructor.");
+        }
+    }
+
+    /**aa
+     * Checks for public static method named 'builder'. Should avoid to use method name, 'builder'.
+     *
+     * @param token the METHOD_DEF AST node
+     */
+    private void checkNonPublicStaticBuilderMethodName(DetailAST token) {
+        DetailAST methodNameToken = token.findFirstToken(TokenTypes.IDENT);
+        if (methodNameToken != null && BUILDER.equals(methodNameToken.getText())) {
+            DetailAST modifiersToken = token.findFirstToken(TokenTypes.MODIFIERS);
+            AccessModifier accessModifier = CheckUtil.getAccessModifierFromModifiersToken(modifiersToken);
+            if (accessModifier.equals(AccessModifier.PUBLIC)
+                && modifiersToken.branchContains(TokenTypes.LITERAL_STATIC)) {
+                log(modifiersToken, "Subclass of ServiceClient should not have a public static method named ''builder''.");
+            }
+        }
+    }
+
+    /**
+     * Checks for the variable field of the subclass of ServiceClient.
+     * These fields should be final because these classes supposed to be immutable class.
+     *
+     * @param token the OBJBLOCK AST node
+     */
+    private void checkClassFieldFinal(DetailAST token) {
+        for (DetailAST ast = token.getFirstChild(); ast != null; ast = ast.getNextSibling()) {
+            if (TokenTypes.VARIABLE_DEF == ast.getType()) {
+                DetailAST modifiersToken = ast.findFirstToken(TokenTypes.MODIFIERS);
+                if (modifiersToken != null && !modifiersToken.branchContains(TokenTypes.FINAL)) {
+                    log(modifiersToken, "The variable field of the subclass of ServiceClient should be final. These classes supposed to be immutable.");
+                }
+            }
+        }
+    }
+
+    /**
+     * Checks for the class name of Service Client. It class should be named <ServiceName>AsyncClient or <ServiceName>Client.
+     *
+     * @param token the LITERAL_CLASS AST node
+     */
+    private void checkServiceClientNaming(DetailAST token) {
+        for (DetailAST ast = token; ast != null; ast = ast.getNextSibling()) {
+            if (TokenTypes.IDENT == ast.getType()) {
+                String className = ast.getText();
+                if (!className.endsWith(ASYNC_CLIENT) && !className.endsWith(CLIENT)) {
```

Comment:
ServiceClient annotation has a field isAsync that will indicate exactly the type of class this should be, so then your log message can be more specific